Add a page for JavaScript prototype pollution attack (#41260)
mdn/content: files/en-us/web/security/attacks/prototype_pollution/index.md

To understand how JavaScript prototypes work, read the following MDN articles:

- [Inheritance and the prototype chain](/en-US/docs/Web/JavaScript/Guide/Inheritance_and_the_prototype_chain)
- [Working with objects](/en-US/docs/Web/JavaScript/Guide/Working_with_objects)

In a nutshell, every object in JavaScript has a `prototype` property which is an object itself and that prototype has its own prototype, too. This is called the JavaScript prototype chain. The chain ends when we reach an object that has `null` for its own prototype. The power of the prototype chain is that it allows an object to inherit properties and methods from its parent. However, a bad practice that JavaScript allows, is that the prototypes of objects can be changed during runtime which makes it possible to install properties at unintended places with malicious values. This even includes the built-in default objects, like the {{jsxref("Object")}} object, which is likely in the prototype chain of the majority of objects in your codebase and therefore makes this attack especially dangerous.

> A key attack vector is the special [`Object.prototype.__proto__`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/proto) property, and how adding a property to `Object.prototype` will make that property accessible on ("polluted to") every object in your application. The `__proto__` property is often used, but you can also reach `Object.prototype` via `yourObject.constructor.prototype`.

Author: I removed "special" and I changed this to have two paragraphs with the goal to explain two things:
1) The _`_proto__` properties are used very often for this attack but, as Josh says, the constructor.prototype property can also be used.
2) Object.prototype is most commonly the target of this. But other prototypes (like Promise.prototype) could also be targets.

```js
Object.prototype.body = "a=1";
Object.prototype.method = "POST";

fetch("https://example.com", {
  mode: "cors",
});
// Promise {status: "pending", body: "a=1", method: "POST"}

// Any new object initialization is now modified to contain additional default properties
console.log({}.method); // "POST"
```

Another dangerous pollution attack target is the {{domxref("HTMLIframeElement.srcdoc")}} property which specifies the content of an {{HTMLElement("iframe")}} element. By overriding its value, it could potentially be possible to execute arbitrary code.

```js
Object.prototype.srcdoc = ["<script>alert(1)<\/script>"];
```

Any configuration objects, like `fetch()`'s {{domxref("RequestInit")}} object in the code example above, or the instantiation of `<iframes>`, or configuration of sanitizers ({{domxref("SanitizerConfig")}} objects) are some of the most sensitive objects and are often targets of prototype pollution attacks.

> 

Reviewer: I think these example are more useful to show how prototypes are malleable in JavaScript (as an illustration of the previous discussion of prototypes being malleable) than they are as examples of attacks.

> 

Author: I agree and I struggle to explain the first step here. I understand how it is game over if you are already able to pollute prototypes, but I don't understand how it usually comes to that. I mean anyone could do `Object.prototype.method = "POST";` in devtools even. Would that make things horrible already? Do pentesters test this way?

I actually try to mention this in the "Vulnerable sources" section. So there I mention URL query strings which https://github.com/BlackFan/client-side-prototype-pollution seems to be all about. And I mention JSON.parse which apparently also can go wrong if you deal with JSON input coming from untrusted sources.

### Vulnerable sources

In order to pollute objects, the attacker needs a way to add arbitrary properties to prototype objects. This may happen as a consequence of [XSS](/en-US/docs/Web/Security/Attacks/XSS), in which the attacker gains direct access to the page's JavaScript execution environment. However, this can also happen without code injection, such as by constructing a payload that contains the `__proto__` property key will later get merged into another object via [spreading](/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax), [`for...in` loops](/en-US/docs/Web/JavaScript/Reference/Statements/for...in), etc., turning it into an assignment operation and therefore triggering the setter.

In a [very common prototype pollution vulnerability](https://github.com/BlackFan/client-side-prototype-pollution), the attacker pollutes query strings in the URL, or alternatively pollution could happen via JSON input, because {{jsxref("JSON.parse()")}} also includes strings like `__proto__`.

## Defenses against prototype pollution

> There are a few options to mitigating prototype pollution vulnerabilities. This section presents them.

Author: I don't know the answer. I can try to guess. Would be good to hear from a SME.

For objects you create and fully control:
- Evaluate if an object is needed or if a Map or Set would be the better choice
- If you're forced to create an object, use Object.create(null), especially for sensitive objects like FetchInit or SanitizerConfig
- If you can't create an object with Object.create(null), consider freezing your object, or make it immutable using a library like Immer
- If you need to create complex and unfrozen objects with a prototype, consider validating your object structure with a schema and reject unneeded properties by setting `additionalProperties` to `false`.

For built-in and third party objects:
- Consider freezing built-in and third party objects, for example by using the [SES](https://github.com/endojs/endo/tree/master/packages/ses#ses) shim.

Runtime defenses:
- Use `--disable-proto` in Node.js to disable `Object.prototype.__proto__`.
- Use `delete Object.prototype.__proto__` in non-Node environments.

---

> title: Prototype pollution

Reviewer 2: 1. Maybe 

```suggestion
title: JavaScript prototype pollution
```

2. Don't forget to link from parent page and put in sidebar

slug: Web/Security/Attacks/Prototype_pollution
page-type: guide
sidebar: security
---

**JavaScript prototype pollution** is a vulnerability where an attacker can add or modify properties on an object's prototype. This means malicious values can unexpectedly appear on every object in your application, often leading to logic errors or additional attacks like [cross-site scripting (XSS)](/en-US/docs/Web/Security/Attacks/XSS).

> To understand how prototypes can be polluted, the following MDN articles are useful resources to get an understanding of how JavaScript prototypes work:

Reviewer 3: ```suggestion
To understand how JavaScript prototypes work, read the following MDN articles:
```

- [Inheritance and the prototype chain](/en-US/docs/Web/JavaScript/Guide/Inheritance_and_the_prototype_chain)
- [Working with objects](/en-US/docs/Web/JavaScript/Guide/Working_with_objects)

> The key thing to understand is the special meaning of the [`Object.prototype.__proto__`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/proto) property, and how adding a property to `Object.prototype` will make that property accessible on ("polluted to") every object in your application.

Reviewer 3: ```suggestion
A key attack vector is the special [`Object.prototype.__proto__`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/proto) property, and how adding a property to `Object.prototype` will make that property accessible on ("polluted to") every object in your application.
```

> Not all prototype pollution involves adding properties to `Object.prototype`, though. Other prototypes can also be polluted, like when overriding {{jsxref("Promise.prototype.then")}} which would be called for every `await` operation, and more.

Reviewer 3: ```suggestion
Not all prototype pollution involves adding properties to `Object.prototype`, though. Other prototypes can also be polluted, like overriding {{jsxref("Promise.prototype.then")}}, which would be called for every `await` operation, and more.
```

> To see the effect of prototype pollution, we can look at the how the following {{domxref("fetch()")}} call can be changed completely. By default, it is a {{HTTPMethod("GET")}} request, but because we polluted the `Object` object's prototype with two new default properties, the `fetch()` call is now transformed into a {{HTTPMethod("POST")}} request.

Author: I added another example that Aaron suggested on `srcdoc` to make things more spooky.

```js
Object.prototype.body = "a=1";
Object.prototype.method = "POST";

fetch("https://example.com", {
  mode: "cors",
});
// Promise {status: "pending", body: "a=1", method: "POST"}

// Any new object initialization is now modified to contain additional default properties
console.log({}.method); // "POST"
```

## Example scenarios

> In order to pollute objects, the attacker needs a way in to add arbitrary properties to prototype objects. Typically, happens in form of some payload that contains the `__proto__` property key will later get merged into another object via [spreading](/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax), [`for...in` loops](/en-US/docs/Web/JavaScript/Reference/Statements/for...in), etc, turning it into an assignment operation and therefore triggering the setter.

Reviewer 3: ```suggestion
In order to pollute objects, the attacker needs a way in to add arbitrary properties to prototype objects. This may happen as a consequence of [XSS](/en-US/docs/Web/Security/Attacks/XSS), in which the attacker gains direct access to the page's JavaScript execution environment. However, this can also happen without code injection, such as by constructing a payload that contains the `__proto__` property key will later get merged into another object via [spreading](/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax), [`for...in` loops](/en-US/docs/Web/JavaScript/Reference/Statements/for...in), etc., turning it into an assignment operation and therefore triggering the setter.
```

Reviewer 2: Yes, but perhaps "way to" not "way in to" in the first sentence.

### Pollution via URL query strings

> In a very common prototype pollution vulnerability, the attacker pollutes query strings in the URL:

Reviewer 2: Are there other common vulnerabilities - perhaps as links?

See also [Client-Side Prototype Pollution](https://github.com/BlackFan/client-side-prototype-pollution) for libraries vulnerable due to {{domxref("document.location")}} parsing and the various URL payloads used.

If you parse the query strings into key-value pairs and treat `__proto__` like an arbitrary string, you risk that `__proto__` will later be merged into a target object like this:

```js
const result = { ...defaultValues, ...query };
```

which you may expect to result in the following object structure for `result`:

```js
{
  existingProperty: "hello",
  __proto__: {
      test: "polluted"
  }
}
```

However, this is not what is happening! Instead, this adds `test` to the object's prototype, which is {{jsxref("Object.prototype")}}. Now all objects in the JavaScript runtime will inherit the `test` property and we have successfully polluted the prototype. The `test` property is likely not meaningful in this case, but other property names might become a problem for your code or for any code in your imported libraries as it could override existing properties or functions.

```js
result.test; // "polluted"
{}.test; // "polluted"
"".test; // "polluted"
[].test; // "polluted"

> // ...

Reviewer 3: Ummm... This is wrong. In this case you constructed a single object, `result`, whose prototype is `{ test: "polluted" }`. You did not successfully taint every object (although this may still be an issue). In order to do that, you need to have code that assigns properties, where the object being assigned could turn out to be `Object.prototype`:

```js
result[x][y] = z;
```

In this case, if `result` is a plain object, then you can set `x = "__proto__"` and `y = "test"` to successfully inject it.

```

## Defenses against prototype pollution

There are a few options to mitigating prototype pollution vulnerabilities. This section presents them.

> ### Create a JavaScript object without a prototype

Author: Yeah, I tried to make it more clear that the default behavior is not a null-prototype unfortunately.

I moved the Map and Set mitigation below Object.freeze now, so it is second after Object.create(null). I hope that makes more sense.

The {{jsxref("Object.create()", "Object.create(null)")}} function allows you to create a new empty object without a prototype. Because of this, there won't be any prototype pollution issues:

```js
let obj = Object.create(null);
obj.__proto__; // undefined
obj.constructor; // undefined
obj["__proto__"]["a"] = 1; // TypeError: Cannot set properties of undefined
```

### Use `Map` and `Set` instead

Instead of using JavaScript objects, use {{jsxref("Map")}} or {{jsxref("Set")}} objects. These are more modern data structures not vulnerable to prototype pollution issues.

> ### Freezing the prototype

Author: Object.seal and Object.preventExtensions are both mentioned in this section now. Object.freeze is presented as the tightest option

You can also freeze the prototype of the built-in base {{jsxref("Object")}}. However, note that in your code or in the code of your dependencies you may want to add to the built-in {{jsxref("Object")}} prototype or other built-in objects (for example, to provide a {{glossary("Polyfill")}} implementation). Given assignment on frozen objects fails silently, it may be difficult to debug.

```js
Object.freeze(Object.prototype);
let obj = {};
obj["__proto__"]["a"] = 1; // fails silently in non-strict mode
obj.a; // undefined
```

You can enable JavaScript's [strict mode](/en-US/docs/Web/JavaScript/Reference/Strict_mode) to see errors when trying to define properties on objects that are not extensible.

Another caveat to note with {{jsxref("Object.freeze()")}} is that it doesn't provide a deep freeze by default. If you want true immutability, you need to recursively freeze every property ([example](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze#deep_freezing)).

```js
const config = deepFreeze({
  url: "https://api.example.org",
  headers: {
    Authorization: "some_token",
  },
});

config.headers.Authorization = "hacked_token"; // fails thanks to deepFreeze
```

> The `deepFreeze` approach is not handling circular references by default, so to create true immutable objects, you may want to look into using a library like [Immer](https://immerjs.github.io/immer/). You could also consider the [SES](https://github.com/endojs/endo/tree/master/packages/ses#ses) shim for [Hardened JavaScript](https://hardenedjs.org) which is a [draft proposal for ECMAScript](https://github.com/tc39/proposal-ses) to freeze every built-in object.

Reviewer 3: SES is first a library and then a proposal, so I would just mention the library at this point because it's unlikely for the proposal to make much progress.

### JSON schema validation

JSON schema validators, such as [ajv](https://ajv.js.org), ensure that the JSON data structure contains the appropriate properties with the appropriate types. To mitigate the prototype pollution attack, reject unneeded properties by setting `additionalProperties` to `false` in the schema.

### Node.js flag `--disable-proto`

> If you are in a Node.js environment, you can disable `Object.prototype.__proto__` with the `--disable-proto=MODE` option where `MODE` is either `delete` (the property is removed entirely), or `throw` (accesses to the property throws an exception with the code `ERR_PROTO_ACCESS`).

Reviewer 3: A simple trick for non-Node is `delete Object.prototype.__proto__`

Author: Is this a trick or advice we should give?

## Defense summary checklist

- Create empty objects with `Object.create(null)`.
- Use `Map` and `Set` objects instead.
- Consider freezing your own objects, for example with a library like [Immer](https://immerjs.github.io/immer/).
- Consider freezing built-in objects, for example by using the [SES](https://github.com/endojs/endo/tree/master/packages/ses#ses) shim.
- Validate your object structure with a schema.

> - Use `--disable-proto` in Node.js.

Reviewer 2: Same as above "what object". I assume disabling proto etc only disables proto on your code? How might it affect imported libraries etc?

Author: Addded "to disable `Object.prototype.__proto__`." because as I understand it, it just disables the special `__proto__` property.
